File: saral_hr/saral_hr/report/esi_register/esi_register.js
// Copyright (c) 2026, sj and contributors
// For license information, please see license.txt

frappe.query_reports["ESI Register"] = {
    filters: [
        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: "\n2023\n2024\n2025\n2026\n2027",
            reqd: 1
        },
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: "\nJanuary\nFebruary\nMarch\nApril\nMay\nJune\nJuly\nAugust\nSeptember\nOctober\nNovember\nDecember",
            reqd: 1
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "MultiSelectList",
            reqd: 1,
            get_data: function(txt) {
                return frappe.db.get_link_options("Company", txt);
            }
        },
        {
            fieldname: "category",
            label: __("Category"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                return frappe.db.get_link_options("Category", txt);
            }
        },
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                let companies = frappe.query_report.get_filter_value("company") || [];
                let result = [];
                frappe.call({
                    method: "saral_hr.saral_hr.report.esi_register.esi_register.get_esi_employees_for_filter",
                    args: {
                        companies: JSON.stringify(companies),
                        txt: txt || ""
                    },
                    async: false,
                    callback: function(r) {
                        result = (r.message || []).map(emp => ({
                            value: emp.employee,
                            description: emp.employee_name
                        }));
                    }
                });
                return result;
            }
        },
    ],

    onload: function(report) {
        frappe.after_ajax(() => {
            report.page.set_primary_action(__("Print"), () => {
                esi_custom_print(report);
            }, "printer");
        });
    }
};

function esi_custom_print(report) {
    let data    = frappe.query_report.data || [];
    let filters = frappe.query_report.get_values() || {};

    let month = filters.month || "";
    let year  = filters.year  || "";

    let rows       = data.filter(r => r.employee_name !== "Total");
    let totals_row = data.find(r => r.employee_name === "Total") || {};

    if (!rows.length) {
        frappe.msgprint(__("No data to print."));
        return;
    }

    let tbody = rows.map((row, idx) => `
        <tr>
            <td class="center">${idx + 1}</td>
            <td class="center">${row.esic_number || "-"}</td>
            <td>${row.employee_id || ""}</td>
            <td>${row.employee_name || ""}</td>
            <td class="center">${fmt(row.days_paid)}</td>
            <td class="num">${fmt(row.gross_salary)}</td>
            <td class="num">${fmt(row.total_esi)}</td>
            <td class="center">${row.date_of_joining || "-"}</td>
            <td class="center">${row.date_of_birth || "-"}</td>
        </tr>
    `).join("");

    tbody += `
        <tr class="total-row">
            <td colspan="5" style="text-align:right; padding-right:12px; font-weight:bold;">Total</td>
            <td class="num">${fmt(totals_row.gross_salary)}</td>
            <td class="num">${fmt(totals_row.total_esi)}</td>
            <td colspan="2"></td>
        </tr>
    `;

    let html = `
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title>ESI Register – ${month} ${year}</title>
            <style>
                * { margin:0; padding:0; box-sizing:border-box; }
                body { font-family: Arial, sans-serif; font-size: 11px; color: #000; }
                .container { padding: 24px; }
                .header { text-align: center; margin-bottom: 16px; }
                .header .company-name { font-size: 16px; font-weight: bold; text-transform: uppercase; }
                .header .report-title { font-size: 13px; font-weight: bold; margin-top: 4px; }
                .header .period { font-size: 11px; margin-top: 2px; color: #444; }
                table { width: 100%; border-collapse: collapse; margin-top: 12px; }
                thead tr th { background-color: #e8e8e8; border: 1px solid #999; padding: 7px 8px; text-align: center; font-size: 11px; font-weight: bold; white-space: nowrap; }
                tbody tr td { border: 1px solid #bbb; padding: 6px 8px; font-size: 11px; white-space: nowrap; }
                tbody tr:nth-child(even):not(.total-row) { background-color: #f9f9f9; }
                td.num { text-align: right; }
                td.center { text-align: center; }
                tr.total-row td { font-weight: bold; background-color: #eef4fb; border-top: 2px solid #555; border-bottom: 2px solid #555; }
                .footer { margin-top: 40px; display: flex; justify-content: space-between; }
                .sign-block { text-align: center; width: 180px; }
                .sign-block .line { border-top: 1px solid #000; margin-bottom: 4px; }
                .sign-block .label { font-size: 10px; color: #444; }
                @media print {
                    @page { size: A4 landscape; margin: 15mm; }
                    body { -webkit-print-color-adjust: exact; print-color-adjust: exact; }
                }
            </style>
        </head>
        <body>
        <div class="container">
            <div class="header">
                <div class="company-name">${frappe.boot.sysdefaults.company || ""}</div>
                <div class="report-title">ESI Register</div>
                <div class="period">For the Month of ${month} ${year}</div>
            </div>
            <table>
                <thead>
                    <tr>
                        <th style="width:45px;">Sr. No.</th>
                        <th style="width:110px;">ESI Number</th>
                        <th style="width:100px;">Employee ID</th>
                        <th style="width:170px;">Employee Name</th>
                        <th style="width:70px;">Days Paid</th>
                        <th style="width:110px;">Gross Salary</th>
                        <th style="width:120px;">ESI Contribution</th>
                        <th style="width:90px;">Join Date</th>
                        <th style="width:90px;">Birth Date</th>
                    </tr>
                </thead>
                <tbody>${tbody}</tbody>
            </table>
            <div class="footer">
                <div class="sign-block"><div class="line"></div><div class="label">Prepared By</div></div>
                <div class="sign-block"><div class="line"></div><div class="label">Checked By</div></div>
                <div class="sign-block"><div class="line"></div><div class="label">Authorised Signatory</div></div>
            </div>
        </div>
        </body>
        </html>
    `;

    let w = window.open("", "_blank");
    w.document.write(html);
    w.document.close();
    w.focus();
    setTimeout(() => { w.print(); }, 500);
}

function fmt(val) {
    return parseFloat(val || 0).toFixed(2);
}

File: saral_hr/saral_hr/report/esi_register/esi_register.py
import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {"fieldname": "esic_number",    "label": _("ESI Number"),      "fieldtype": "Data",  "width": 150},
        {"fieldname": "employee_id",    "label": _("Employee ID"),      "fieldtype": "Data",  "width": 120},
        {"fieldname": "employee_name",  "label": _("Employee Name"),    "fieldtype": "Data",  "width": 200},
        {"fieldname": "days_paid",      "label": _("Days Paid"),        "fieldtype": "Float", "precision": 2, "width": 100},
        {"fieldname": "gross_salary",   "label": _("Gross Salary"),     "fieldtype": "Float", "precision": 2, "width": 140},
        {"fieldname": "total_esi",      "label": _("ESI Contribution"), "fieldtype": "Float", "precision": 2, "width": 140},
        {"fieldname": "date_of_joining","label": _("Join Date"),        "fieldtype": "Date",  "width": 110},
        {"fieldname": "date_of_birth",  "label": _("Birth Date"),       "fieldtype": "Date",  "width": 110},
    ]


def get_conditions(filters):
    conditions = ""

    if filters.get("company"):
        companies = frappe.parse_json(filters.get("company"))
        filters["companies"] = tuple(companies)
        conditions += " AND ss.company IN %(companies)s"

    if filters.get("employee"):
        employees = frappe.parse_json(filters.get("employee"))
        filters["employees"] = tuple(employees)
        conditions += " AND ss.employee IN %(employees)s"

    if filters.get("year") and filters.get("month"):
        month_map = {
            "January": 1, "February": 2, "March": 3, "April": 4,
            "May": 5, "June": 6, "July": 7, "August": 8,
            "September": 9, "October": 10, "November": 11, "December": 12
        }
        month_num = month_map.get(filters.get("month"))
        if month_num:
            start_date = "{}-{:02d}-01".format(filters.get("year"), month_num)
            filters["start_date"] = start_date
            conditions += " AND ss.start_date = %(start_date)s"

    return conditions


def get_data(filters):
    if not filters.get("year") or not filters.get("month"):
        return []
    if not filters.get("company"):
        return []

    conditions = get_conditions(filters)

    # Category filter — Company Link se INNER JOIN
    category_join = ""
    if filters.get("category"):
        categories = frappe.parse_json(filters.get("category"))
        if categories:
            category_join = "INNER JOIN `tabCompany Link` cl_cat ON cl_cat.employee = ss.employee"
            filters["categories"] = tuple(categories)
            conditions += " AND cl_cat.category IN %(categories)s"

    salary_slips = frappe.db.sql(
        f"""
        SELECT
            ss.name            AS salary_slip,
            ss.employee        AS employee,
            ss.employee_name   AS employee_name,
            ss.payment_days    AS days_paid,
            ss.total_earnings  AS gross_salary,
            emp.name           AS employee_id,
            emp.date_of_joining,
            emp.date_of_birth,
            emp.esic_number
        FROM
            `tabSalary Slip` ss
        LEFT JOIN
            `tabEmployee` emp ON emp.name = ss.employee
        {category_join}
        WHERE
            ss.docstatus = 1
            {conditions}
        ORDER BY
            ss.employee_name
        """,
        filters,
        as_dict=1,
    )

    if not salary_slips:
        return []

    slip_names = [d.salary_slip for d in salary_slips]

    esi_rows = frappe.db.sql(
        """
        SELECT sd.parent AS salary_slip, sd.amount
        FROM `tabSalary Details` sd
        WHERE
            sd.parent IN %(slip_names)s
            AND sd.parentfield = 'deductions'
            AND LOWER(sd.salary_component) LIKE '%%esic%%'
            AND sd.amount > 0
        """,
        {"slip_names": tuple(slip_names)},
        as_dict=1,
    )

    esi_map = {}
    for row in esi_rows:
        esi_map[row.salary_slip] = esi_map.get(row.salary_slip, 0) + flt(row.amount)

    data = []
    grand_gross = 0
    grand_esi   = 0

    for slip in salary_slips:
        total_esi = flt(esi_map.get(slip.salary_slip, 0))
        if total_esi <= 0:
            continue

        grand_gross += flt(slip.gross_salary, 2)
        grand_esi   += total_esi

        data.append({
            "esic_number":     slip.esic_number or "-",
            "employee_id":     slip.employee_id or slip.employee,
            "employee_name":   slip.employee_name,
            "days_paid":       flt(slip.days_paid, 2),
            "gross_salary":    flt(slip.gross_salary, 2),
            "total_esi":       total_esi,
            "date_of_joining": slip.date_of_joining,
            "date_of_birth":   slip.date_of_birth,
        })

    if data:
        data.append({
            "esic_number":     "",
            "employee_id":     "",
            "employee_name":   "Total",
            "days_paid":       "",
            "gross_salary":    flt(grand_gross, 2),
            "total_esi":       flt(grand_esi, 2),
            "date_of_joining": "",
            "date_of_birth":   "",
            "bold": 1,
        })

    return data


@frappe.whitelist()
def get_esi_employees_for_filter(companies=None, txt=""):
    companies = frappe.parse_json(companies) if companies else []
    company_condition = ""
    params = {"txt": f"%{txt}%"}

    if companies:
        company_condition = "AND ss.company IN %(companies)s"
        params["companies"] = tuple(companies)

    results = frappe.db.sql("""
        SELECT DISTINCT
            ss.employee       AS employee,
            ss.employee_name  AS employee_name
        FROM `tabSalary Slip` ss
        INNER JOIN `tabSalary Details` sd
            ON  sd.parent      = ss.name
            AND sd.parentfield = 'deductions'
            AND LOWER(sd.salary_component) LIKE '%%esic%%'
            AND sd.amount > 0
        WHERE
            ss.docstatus = 1
            AND (ss.employee LIKE %(txt)s OR ss.employee_name LIKE %(txt)s)
            {company_condition}
        ORDER BY ss.employee_name
        LIMIT 50
    """.format(company_condition=company_condition), params, as_dict=1)

    return results